Git tip fails to build for me on Fedora 24 (using Rust 1.10.0)

I'm trying to build Rust and Cargo from source because I want to put them in a non-standard (non-system) location, and building Cargo fails on my 64-bit Fedora 24 system with an error in the 'make' phase. Here is a full transcript of the make (with the prequel steps included because why not, it reproduces easily).

```
$ cd /tmp
$ git clone https://github.com/rust-lang/cargo
$ cd cargo
$ ./configure
[nattering deleted]
$ make
/usr/bin/python2.7 /tmp/cargo//src/etc/dl-snapshot.py x86_64-unknown-linux-gnu
running: curl -o target/dl/cargo-nightly-x86_64-unknown-linux-gnu.tar.gz https://static.rust-lang.org/cargo-dist/2016-03-21/cargo-nightly-x86_64-unknown-linux-gnu.tar.gz
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/manifest.in
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/bin/cargo
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/share/zsh/site-functions/_cargo
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/share/man/man1/cargo.1
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/share/doc/cargo/LICENSE-THIRD-PARTY
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/share/doc/cargo/LICENSE-APACHE
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/share/doc/cargo/LICENSE-MIT
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/share/doc/cargo/README.md
extracting cargo-nightly-x86_64-unknown-linux-gnu/cargo/etc/bash_completion.d/cargo
touch target/snapshot/bin/cargo
/u/cks/bin/bin.i386-linux/rustc -V
rustc 1.10.0-dev
target/snapshot/bin/cargo --version
cargo 0.10.0-nightly (132b82d 2016-03-19)
target/snapshot/bin/cargo build --target x86_64-unknown-linux-gnu --manifest-path /tmp/cargo//Cargo.toml --release   
error: unexpected character in cfg `, expected parens, a comma, an identifier, or a string
Makefile:104: recipe for target 'cargo-x86_64-unknown-linux-gnu' failed
make: *** [cargo-x86_64-unknown-linux-gnu] Error 101
```
On further testing, this appears to be something in my $HOME that Cargo doesn't like. It is not telling me what, and it is not $HOME/.cargo (because I just deleted that entire directory hierarchy). strace is not particularly revealing about what what file or configuration cargo dislikes.

--- a/src/cargo/core/compiler/build_context/target_info.rs
+++ b/src/cargo/core/compiler/build_context/target_info.rs
@@ -170,7 +170,15 @@ impl TargetInfo {
             }
         };
 
-        let cfg = lines.map(Cfg::from_str).collect::<CargoResult<Vec<_>>>()?;
+        let cfg = lines
+            .map(Cfg::from_str)
+            .collect::<CargoResult<Vec<_>>>()
+            .chain_err(|| {
+                format!(
+                    "failed to parse the cfg from `rustc --print=cfg`, got:\n{}",
+                    output
+                )
+            })?;
 
         Ok(TargetInfo {
             crate_type_process,

--- a/src/cargo/util/rustc.rs
+++ b/src/cargo/util/rustc.rs
@@ -52,7 +52,12 @@ impl Rustc {
                 .lines()
                 .find(|l| l.starts_with("host: "))
                 .map(|l| &l[6..])
-                .ok_or_else(|| internal("rustc -v didn't have a line for `host:`"))?;
+                .ok_or_else(|| {
+                    failure::format_err!(
+                        "`rustc -vV` didn't have a line for `host:`, got:\n{}",
+                        verbose_version
+                    )
+                })?;
             triple.to_string()
         };
 

--- a/tests/testsuite/cfg.rs
+++ b/tests/testsuite/cfg.rs
@@ -446,3 +446,138 @@ fn cfg_looks_at_rustflags_for_target() {
         .env("RUSTFLAGS", "--cfg with_b")
         .run();
 }
+
+#[cargo_test]
+fn bad_cfg_discovery() {
+    // Check error messages when `rustc -v` and `rustc --print=*` parsing fails.
+    //
+    // This is a `rustc` replacement which behaves differently based on an
+    // environment variable.
+    let p = project()
+        .at("compiler")
+        .file("Cargo.toml", &basic_manifest("compiler", "0.1.0"))
+        .file(
+            "src/main.rs",
+            r#"
+fn run_rustc() -> String {
+    let mut cmd = std::process::Command::new("rustc");
+    for arg in std::env::args_os().skip(1) {
+        cmd.arg(arg);
+    }
+    String::from_utf8(cmd.output().unwrap().stdout).unwrap()
+}
+
+fn main() {
+    let mode = std::env::var("FUNKY_MODE").unwrap();
+    if mode == "bad-version" {
+        println!("foo");
+        return;
+    }
+    if std::env::args_os().any(|a| a == "-vV") {
+        print!("{}", run_rustc());
+        return;
+    }
+    if mode == "no-crate-types" {
+        return;
+    }
+    if mode == "bad-crate-type" {
+        println!("foo");
+        return;
+    }
+    let output = run_rustc();
+    let mut lines = output.lines();
+    let sysroot = loop {
+        let line = lines.next().unwrap();
+        if line.contains("___") {
+            println!("{}", line);
+        } else {
+            break line;
+        }
+    };
+    if mode == "no-sysroot" {
+        return;
+    }
+    println!("{}", sysroot);
+    if mode != "bad-cfg" {
+        panic!("unexpected");
+    }
+    println!("123");
+}
+"#,
+        )
+        .build();
+    p.cargo("build").run();
+    let funky_rustc = p.bin("compiler");
+
+    let p = project().file("src/lib.rs", "").build();
+
+    p.cargo("build")
+        .env("RUSTC", &funky_rustc)
+        .env("FUNKY_MODE", "bad-version")
+        .with_status(101)
+        .with_stderr(
+            "\
+[ERROR] `rustc -vV` didn't have a line for `host:`, got:
+foo
+
+",
+        )
+        .run();
+
+    p.cargo("build")
+        .env("RUSTC", &funky_rustc)
+        .env("FUNKY_MODE", "no-crate-types")
+        .with_status(101)
+        .with_stderr(
+            "\
+[ERROR] malformed output when learning about crate-type bin information
+command was: `[..]compiler[..] --crate-name ___ [..]`
+(no output received)
+",
+        )
+        .run();
+
+    p.cargo("build")
+        .env("RUSTC", &funky_rustc)
+        .env("FUNKY_MODE", "no-sysroot")
+        .with_status(101)
+        .with_stderr(
+            "\
+[ERROR] output of --print=sysroot missing when learning about target-specific information from rustc
+command was: `[..]compiler[..]--crate-type [..]`
+
+--- stdout
+[..]___[..]
+[..]___[..]
+[..]___[..]
+[..]___[..]
+[..]___[..]
+[..]___[..]
+
+",
+        )
+        .run();
+
+    p.cargo("build")
+        .env("RUSTC", &funky_rustc)
+        .env("FUNKY_MODE", "bad-cfg")
+        .with_status(101)
+        .with_stderr(
+            "\
+[ERROR] failed to parse the cfg from `rustc --print=cfg`, got:
+[..]___[..]
+[..]___[..]
+[..]___[..]
+[..]___[..]
+[..]___[..]
+[..]___[..]
+[..]
+123
+
+
+Caused by:
+  unexpected character in cfg `1`, expected parens, a comma, an identifier, or a string
+",
+        )
+        .run();
+}

